From 9e1f7cb8dea3440a75f5ba2559614be11245e600 Mon Sep 17 00:00:00 2001
From: Christoph Hellwig <hch@lst.de>
Date: Fri, 23 Jul 2021 15:50:17 -0700
Subject: [PATCH 246/263] [Backport] mm: call flush_dcache_page() in
 memcpy_to_page() and memzero_page()

mainline inclusion
from mainline-v5.14-rc3
commit 8dad53a11f8d94dceb540a5f8f153484f42be84b
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6AVM6

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=8dad53a11f8d94dceb540a5f8f153484f42be84b

--------------------------------

memcpy_to_page and memzero_page can write to arbitrary pages, which
could be in the page cache or in high memory, so call
flush_kernel_dcache_pages to flush the dcache.

This is a problem when using these helpers on dcache challeneged
architectures.  Right now there are just a few users, chances are no one
used the PC floppy driver, the aha1542 driver for an ISA SCSI HBA, and a
few advanced and optional btrfs and ext4 features on those platforms yet
since the conversion.

Link: https://lkml.kernel.org/r/20210713055231.137602-2-hch@lst.de
Fixes: bb90d4bc7b6a ("mm/highmem: Lift memcpy_[to|from]_page to core")
Fixes: 28961998f858 ("iov_iter: lift memzero_page() to highmem.h")
Signed-off-by: Christoph Hellwig <hch@lst.de>
Reviewed-by: Ira Weiny <ira.weiny@intel.com>
Cc: Chaitanya Kulkarni <chaitanya.kulkarni@wdc.com>
Cc: <stable@vger.kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Conflicts:
	include/linux/highmem.h
Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/highmem.h | 1 +
 lib/iov_iter.c          | 1 +
 2 files changed, 2 insertions(+)

diff --git a/include/linux/highmem.h b/include/linux/highmem.h
index cc5fe6c620ad..1ad238518f1b 100644
--- a/include/linux/highmem.h
+++ b/include/linux/highmem.h
@@ -384,6 +384,7 @@ static inline void memcpy_from_page(char *to, struct page *page,
 	char *from = kmap_atomic(page);
 
 	memcpy(to, from + offset, len);
+	flush_dcache_page(page);
 	kunmap_atomic(from);
 }
 
diff --git a/lib/iov_iter.c b/lib/iov_iter.c
index 0a4b7aa47097..153f915e3fb5 100644
--- a/lib/iov_iter.c
+++ b/lib/iov_iter.c
@@ -471,6 +471,7 @@ static void memzero_page(struct page *page, size_t offset, size_t len)
 {
 	char *addr = kmap_atomic(page);
 	memset(addr + offset, 0, len);
+	flush_dcache_page(page);
 	kunmap_atomic(addr);
 }
 
-- 
2.17.1

